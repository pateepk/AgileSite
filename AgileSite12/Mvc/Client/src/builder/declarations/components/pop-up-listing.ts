export interface PopupListingElement {
  readonly name: string;
  readonly iconClass: string;
  readonly key: string;
  readonly description: string;
}
