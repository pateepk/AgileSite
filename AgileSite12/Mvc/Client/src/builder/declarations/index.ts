export * from "./api";
export * from "./config";
export * from "./entities";
export * from "./events";
export * from "./logger";
export * from "./services";
export * from "./store";
export * from "./components";
export * from "./modal-dialog";
