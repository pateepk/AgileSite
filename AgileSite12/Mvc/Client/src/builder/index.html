<!DOCTYPE html>
<html dir="ltr" lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0">
  <title>Stencil Component Starter</title>
  <script src="/build/components.js"></script>
  <style>
    .devider {
      margin-top: 1em;
    }
  </style>
</head>

<body>

  <kentico-add-component-button zone-identifier="identifier"></kentico-add-component-button>
  <div class="devider">
  <kentico-pop-up-container widgets="[]"></kentico-pop-up-container>
</body>

</html>
