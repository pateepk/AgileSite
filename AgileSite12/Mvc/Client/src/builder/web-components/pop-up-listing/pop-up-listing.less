:host {
  @import 'src/builder/assets/styles/icon-font.less';
  @import 'src/builder/assets/styles/generated/icon-classes.less';
  @import "src/builder/assets/styles/shared.less";

  @pop-up-item-height: 2.5 * @base-unit;
  @pop-up-item-icon-size: 1.5 * @base-unit;
  @pop-up-item-icon-padding: 0.5 * @base-unit;
  @pop-up-item-width-double-column: 0.5 * @pop-up-container-width - @pop-up-padding;
  @pop-up-item-width-single-column: @pop-up-container-width - @pop-up-padding;
  @pop-up-item-label-width-double-column: @pop-up-item-width-double-column - @pop-up-item-icon-size - (2 * @pop-up-item-icon-padding) - @pop-up-padding;
  @pop-up-item-label-width-single-column: @pop-up-item-width-single-column - @pop-up-item-icon-size - (2 * @pop-up-item-icon-padding) - @pop-up-padding;

  .ktc-pop-up-item-listing {
    background-color: @color-white;
    padding-top: @pop-up-padding;
    padding-bottom: @pop-up-padding;

    &.ktc-pop-up-item-listing--empty {
      padding: 0.5 * @base-unit;
    }

    ul {
      margin: 0;
      padding: 0;

      .ktc-pop-up-item-single-column {
        width: 100%;

        .ktc-pop-up-item-label {
          width: @pop-up-item-label-width-single-column;
        }

        .ktc-pop-up-item-label--no-icon {
          width: @pop-up-item-width-single-column - 2 * @pop-up-padding;
        }
      }

      .ktc-pop-up-item-double-column {
        width: @pop-up-item-width-double-column;

        .ktc-pop-up-item-label {
          width: @pop-up-item-label-width-double-column;
        }

        .ktc-pop-up-item-label--no-icon {
          width: @pop-up-item-width-double-column - 2 * @pop-up-padding;
        }
      }

      .ktc-pop-up-item {
        vertical-align: top;
        display: inline-block;
        height: @pop-up-item-height;

        * {
          cursor: pointer;
        }

        &:hover {
          background-color: @color-blue-130;
        }

        &.ktc-pop-up-item--active {
          background-color: @color-blue-130;

          &, * {
            pointer-events: none;
          }
        }

        .ktc-pop-up-item-icon {
          box-sizing: content-box;
          width: @pop-up-item-icon-size;
          height: @pop-up-item-icon-size;
          padding: @pop-up-item-icon-padding;
          float: left;

          > i {
            font-size: @icon-size-100;
          }
        }

        .ktc-pop-up-item-label {
          height: @pop-up-item-height;
          display: inline-block;
          line-height: @pop-up-item-height;
          font-size: 0.875 * @base-unit;
          text-overflow: ellipsis;
          overflow: hidden;
          white-space: nowrap;
          padding: 0 @pop-up-padding;
        }

        .ktc-pop-up-item-icon ~ .ktc-pop-up-item-label {
          padding: 0;
        }
      }
    }
  }
}
