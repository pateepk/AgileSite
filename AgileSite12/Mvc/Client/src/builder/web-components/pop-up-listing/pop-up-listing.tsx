import { Component, Event, EventEmitter, Prop } from "@stencil/core";

import { PopupListingElement } from "@/builder/declarations/index";

@Component({
  tag: "kentico-pop-up-listing",
  styleUrl: "pop-up-listing.less",
  shadow: true,
})
export class PopupListing {

  @Prop() items: PopupListingElement[] = [];
  @Prop() activeItemIdentifier: string;
  @Prop() noItemsAvailableMessage: string;
  @Prop() singleColumn: boolean;

  @Event() selectItem: EventEmitter<PopupListingElement>;

  getItemTitle = (item: PopupListingElement): string => {
    return item.description ? `${item.name}\n${item.description}` : item.name;
  }

  onItemClick = (item: PopupListingElement): string => {
    if (item.key === this.activeItemIdentifier) {
      return;
    }

    this.selectItem.emit(item);
  }

  render() {
    const listItem = (item: PopupListingElement) =>
      <li class={{
        "ktc-pop-up-item": true,
        "ktc-pop-up-item--active": item.key === this.activeItemIdentifier,
        "ktc-pop-up-item-double-column": !this.singleColumn,
        "ktc-pop-up-item-single-column": this.singleColumn,
      }}
        title={this.getItemTitle(item)}
        onClick={() => this.onItemClick(item)}
        data-component-identifier={item.key}
      >
        {!item.iconClass ? "" :
          <div class="ktc-pop-up-item-icon">
            <i aria-hidden="true" class={item.iconClass}></i>
          </div>
        }
        <div class={{ "ktc-pop-up-item-label": true, "ktc-pop-up-item-label--no-icon": !item.iconClass }
        }>{item.name}</div>
      </li>;

    return (
      <div class={{ "ktc-pop-up-item-listing": true, "ktc-pop-up-item-listing--empty": this.items.length === 0 }}>
        {this.items.length === 0 ? this.noItemsAvailableMessage :
          <ul>
            {this.items.map(listItem)}
          </ul>
        }
      </div>
    );
  }
}
