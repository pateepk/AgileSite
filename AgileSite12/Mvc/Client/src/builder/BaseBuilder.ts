import { injectable } from "inversify";
import { Store } from "redux";
import Vue, { Component as VueComponent, CreateElement } from "vue";

import { initHttpClient } from "@/builder/api/client";
import { BuilderConfig } from "@/builder/builderConfig";
import { GLOBAL_WRAPPER_ELEMENT_CLASS } from "@/builder/constants";
import { LocalizationService, State } from "@/builder/declarations";
import { initializeDragAndDrop, initializeSectionDragAndDrop } from "@/builder/drag-and-drop";
import { calculateListingOffset, getPosition } from "@/builder/helpers/position";
import { addCancelScreenLockEventListeners } from "@/builder/helpers/screen-lock";
import { logger } from "@/builder/logger";
import { MessagingService } from "@/builder/services/MessagingService";
import { PageConfigurationService } from "@/builder/services/page-configuration/PageConfigurationService";
import { postMessage } from "@/builder/services/post-message";
import { setPopupPosition } from "@/builder/store/pop-up/actions";
import { MessageTypes } from "@/builder/types";

import { createStoreProviderMixin } from "./helpers/connector";
import { registerModalDialogApi } from "./services/modal-dialog";

const ignoredElements = [
  /^kentico-/,
];

Vue.config.ignoredElements = ignoredElements;
// Turns off the 'You are running Vue in development mode...' warning on Vue startup.
Vue.config.productionTip = false;

@injectable()
export abstract class Builder {
  /**
   * Indicates whether the page builder was already initialized.
   */
  private wasInitialized = false;

  public constructor(
    protected readonly localizationService: LocalizationService,
    private readonly pageConfigurationService: PageConfigurationService,
    private readonly messagingService: MessagingService,
  ) { }

  /**
   * Initializes the builder application.
   * @param config Builder configuration object
   * @param configureStore Configuration function for the store.
   * @param areaComponent Initialization area component.
   * @param moveWidgetAction Function used for widget move handling.
   * @param moveSectionAction Function used for section move handling.
   */
  protected async initialize<TState extends State>(
    config: BuilderConfig,
    configureStore: (state: State) => Store<TState>,
    globalComponentWrapper: VueComponent,
    areaComponent: VueComponent,
    moveWidgetAction,
    moveSectionAction,
  ): Promise<Store<TState>> {
    if (this.wasInitialized) {
      return;
    }

    this.wasInitialized = true;

    try {
      config.validate();

      // Provide localization service to all components
      Vue.prototype.$_localizationService = this.localizationService;

      // Need to initialize the client before making any API calls
      initHttpClient(config);

      const pageConfiguration = await this.pageConfigurationService.getConfiguration(config);
      const store = configureStore(pageConfiguration);

      registerModalDialogApi(store.dispatch, this.localizationService);

      this.messagingService.registerListener(store.getState, store.dispatch, config.allowedOrigins);

      // Request displayed variants from session storage
      postMessage(MessageTypes.GET_DISPLAYED_WIDGET_VARIANTS, null, "*");

      // Update widget list position on resize event
      window.addEventListener("resize", () => {
        this.updatePopupContainerPosition(store);
      });

      initializeDragAndDrop(store.dispatch, store.getState, moveWidgetAction);
      this.initializeEditableAreas(store, areaComponent);
      initializeSectionDragAndDrop(store.dispatch, store.getState, moveSectionAction);

      this.initializeGlobalComponentWrapper(store, globalComponentWrapper);
      addCancelScreenLockEventListeners();

      return store;
    } catch (err) {
      logger.logException(err);
    }
  }

  private getGlobalComponentWrapperHost(): HTMLElement {
    return document.body;
  }

  /**
   * Initializes global component wrapper.
   * @param store Redux store.
   */
  private initializeGlobalComponentWrapper(store: Store<State>, globalComponentWrapper: VueComponent) {
    this.createGlobalComponentWrapperHost();
    new Vue({
      mixins: [createStoreProviderMixin(store)],
      render: (h: CreateElement) => h(globalComponentWrapper),
    }).$mount(`.${GLOBAL_WRAPPER_ELEMENT_CLASS}`);
  }

  /**
   * Creates a HTML element to host the GlobalComponentWrapper component.
   */
  private createGlobalComponentWrapperHost() {
    const globalComponentSlot = document.createElement("div");
    globalComponentSlot.classList.add(GLOBAL_WRAPPER_ELEMENT_CLASS);
    this.getGlobalComponentWrapperHost().appendChild(globalComponentSlot);
  }

  /**
   * Updates widget list position according to the browser's viewport.
   * @param store Store with current page configuration.
   */
  private updatePopupContainerPosition(store: Store<State>) {
    const componentIdentifier = store.getState().popup.componentIdentifier;

    if (componentIdentifier !== null) {
      const addComponent: Element = document.querySelector(
        "kentico-add-component-button[is-active='true'], " +
        "kentico-personalization-button[show-personalization-popup]");

      if (addComponent !== null) {
        const clientRect = addComponent.getBoundingClientRect();
        const newPosition = getPosition(clientRect);
        const newOffset = calculateListingOffset(clientRect);

        store.dispatch(setPopupPosition(newPosition, newOffset));
      }
    }
  }

  private initializeEditableAreas(store: Store<State>, areaComponent: VueComponent) {
    const editableAreasIdentifiers = Object.keys(store.getState().editableAreas);

    editableAreasIdentifiers.forEach((areaId) => {
      new Vue({
        mixins: [createStoreProviderMixin(store)],
        render: (h) => h(areaComponent, {
          props: {
            identifier: areaId,
          },
        }),
      }).$mount(`[data-kentico-editable-area-id="${areaId}"]`);
    });
  }
}
