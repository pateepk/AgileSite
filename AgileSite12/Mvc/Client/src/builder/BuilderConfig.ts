import { BuilderConfiguration, BuilderConstants, ConfigurationEndpoints, SelectorsConfig } from "@/builder/declarations";

/**
 * Represents general configuration for a builder object.
 */
export abstract class BuilderConfig implements BuilderConfiguration {
  constructor(readonly applicationPath: string, readonly configurationEndpoints: ConfigurationEndpoints,
              readonly metadataEndpoint: string, readonly allowedOrigins: string[], readonly constants: BuilderConstants,
              readonly selectors: SelectorsConfig) {
  }

  /**
   * Validates the builder configuration.
   * @throws Throws an error when validation fails.
   */
  public abstract validate(): void;
  public abstract getComponentMarkupData(postedData: object): object;
}
