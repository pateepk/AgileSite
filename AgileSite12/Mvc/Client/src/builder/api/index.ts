/**
 * Responsible for communication with the API.
 * @module api
 */

import { builderConfig, http } from "@/builder/api/client";
import { getService } from "@/builder/container";
import { Metadata, NormalizedConfiguration, NormalizerService, State } from "@/builder/declarations";
import { SERVICE_TYPES } from "@/builder/types";

import { configuration, metadata } from "./endpoints";

/**
 * Requests default properties on the specified URL.
 * @param componentDefaultPropertiesUrl Component default properties URL.
 */
const getComponentDefaultProperties = async (componentDefaultPropertiesUrl: string): Promise<object> => {
  const response = await http.get(componentDefaultPropertiesUrl);

  return response.data;
};

/**
 * Requests markup for a component on the specified URL.
 * @param componentMarkupUrl URL for component markup retrieval.
 * @param properties Component properties.
 */
const getComponentMarkup = async (componentMarkupUrl: string, properties: object): Promise<string> => {
  const data = builderConfig.getComponentMarkupData(properties);
  const response = await http.post(componentMarkupUrl, data);

  return response.data;
};

/**
 * Requests metadata for all components.
 */
const getComponentMetadata = async (): Promise<Metadata> => {
  const response = await metadata.get();

  return getService<NormalizerService>(SERVICE_TYPES.NormalizerService).normalizeMetadata(response.data);
};

/**
 * Requests state for the specified page.
 */
const getState = async (): Promise<NormalizedConfiguration> => {
  const response = await configuration.load();

  return getService<NormalizerService>(SERVICE_TYPES.NormalizerService).normalizeConfiguration(response.data);
};

/**
 * Sends a request to save the page's state.
 * @param state State of the page.
 * @param guid Page GUID.
 */
const saveState = async (state: State, guid: string): Promise<any> => {
  const denormalizedState = getService<NormalizerService>(SERVICE_TYPES.NormalizerService).denormalizeConfiguration(state);

  return configuration.save(denormalizedState, guid);
};

export {
  getComponentDefaultProperties,
  getComponentMarkup,
  getComponentMetadata,
  getState,
  saveState,
};
