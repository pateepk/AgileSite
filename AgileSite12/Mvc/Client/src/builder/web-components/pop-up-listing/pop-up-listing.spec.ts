import { TestWindow } from "@stencil/core/testing";

import { PopupListingElement } from "@/builder/declarations";
import { PopupListing } from "./pop-up-listing";

export const POPUP_LISTING_ITEMS: PopupListingElement[] = [
  { name: "NoDescriptionItem1", iconClass: "icon-test", key: "GUID1", description: "" },
  { name: "NoIconItem2", iconClass: null, key: "GUID2", description: "Item2" },
  { name: "NoIconNorDescriptionItem3", iconClass: null, key: "GUID3", description: "" },
  { name: "Item4", iconClass: "icon-test", key: "GUID4", description: "Item4" },
  { name: "Item5", iconClass: "icon-test", key: "GUID5", description: "Item5" },
];

describe("pop-up-listing", () => {
  it("should build", () => {
    expect(new PopupListing()).toBeTruthy();
  });

  describe("getItemTitle", () => {
    it("should return both name and description separated by new line", () => {
      const component = new PopupListing();
      const item = POPUP_LISTING_ITEMS[1];

      const expectedItemTitle = `${item.name}\n${item.description}`;
      const actualItemTitle = component.getItemTitle(item);

      expect(actualItemTitle).toBe(expectedItemTitle);
    });

    it("should return the name of the item", () => {
      const component = new PopupListing();
      const item = POPUP_LISTING_ITEMS[0];

      const expectedItemTitle = item.name;
      const actualItemTitle = component.getItemTitle(item);

      expect(actualItemTitle).toBe(expectedItemTitle);
    });
  });

  describe("onItemClick", () => {
    it("should emit event with item", () => {
      const component = new PopupListing();
      component.items = POPUP_LISTING_ITEMS;

      const selectItemSpy = jest.fn();
      component.selectItem = {
        emit: selectItemSpy,
      };

      component.onItemClick(POPUP_LISTING_ITEMS[0]);
      expect(selectItemSpy).toHaveBeenCalledWith(POPUP_LISTING_ITEMS[0]);
    });

    it("should not emit event when currently active item is passed", () => {
      const component = new PopupListing();
      component.items = POPUP_LISTING_ITEMS;
      component.activeItemIdentifier = POPUP_LISTING_ITEMS[0].key;

      const selectItemSpy = jest.fn();
      component.selectItem = {
        emit: selectItemSpy,
      };

      component.onItemClick(POPUP_LISTING_ITEMS[0]);
      expect(selectItemSpy).not.toHaveBeenCalled();
    });
  });

  describe("rendering", () => {
    let component: HTMLKenticoPopUpListingElement;
    let testWindow: TestWindow;

    beforeEach(async () => {
      testWindow = new TestWindow();
      component = await testWindow.load({
        components: [PopupListing],
        html: "<kentico-pop-up-listing></kentico-pop-up-listing>",
      });
    });

    it("should render with no items available message", async () => {
      component.noItemsAvailableMessage = "TEST-MESSAGE";
      await testWindow.flush();

      expect(component).toMatchSnapshot();
    });

    it("should render with item list", async () => {
      component.items = POPUP_LISTING_ITEMS;
      await testWindow.flush();

      expect(component).toMatchSnapshot();
    });
  });

  describe("item click", () => {
    let component: HTMLKenticoPopUpListingElement;
    let testWindow: TestWindow;

    beforeEach(async () => {
      testWindow = new TestWindow();
      component = await testWindow.load({
        components: [PopupListing],
        html: "<kentico-pop-up-listing></kentico-pop-up-listing>",
      });
      component.items = [ POPUP_LISTING_ITEMS[0] ];
    });

    it("click on normal item should emit event", async () => {
      await testWindow.flush();
      const eventSpy = jest.fn();
      testWindow.document.addEventListener("selectItem", eventSpy);

      component.querySelector("li").click();

      expect(eventSpy).toHaveBeenCalled();
    });

    it("click on active item should not emit event", async () => {
      component.activeItemIdentifier = POPUP_LISTING_ITEMS[0].key;
      await testWindow.flush();
      const eventSpy = jest.fn();
      testWindow.document.addEventListener("selectItem", eventSpy);

      component.querySelector("li").click();

      expect(eventSpy).not.toHaveBeenCalled();
    });
  });
});
