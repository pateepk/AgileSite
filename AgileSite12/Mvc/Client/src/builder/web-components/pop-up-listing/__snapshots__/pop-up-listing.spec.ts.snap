// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pop-up-listing rendering should render with item list 1`] = `
<kentico-pop-up-listing
  class="hydrated"
>
  
  <div
    class="ktc-pop-up-item-listing"
  >
    <ul>
      <li
        class="ktc-pop-up-item ktc-pop-up-item-double-column"
        data-component-identifier="GUID1"
        title="NoDescriptionItem1"
      >
        <div
          class="ktc-pop-up-item-icon"
        >
          <i
            aria-hidden="true"
            class="icon-test"
          />
        </div>
        <div
          class="ktc-pop-up-item-label"
        >
          NoDescriptionItem1
        </div>
      </li>
      <li
        class="ktc-pop-up-item ktc-pop-up-item-double-column"
        data-component-identifier="GUID2"
        title="NoIconItem2
Item2"
      >
        
        <div
          class="ktc-pop-up-item-label ktc-pop-up-item-label--no-icon"
        >
          NoIconItem2
        </div>
      </li>
      <li
        class="ktc-pop-up-item ktc-pop-up-item-double-column"
        data-component-identifier="GUID3"
        title="NoIconNorDescriptionItem3"
      >
        
        <div
          class="ktc-pop-up-item-label ktc-pop-up-item-label--no-icon"
        >
          NoIconNorDescriptionItem3
        </div>
      </li>
      <li
        class="ktc-pop-up-item ktc-pop-up-item-double-column"
        data-component-identifier="GUID4"
        title="Item4
Item4"
      >
        <div
          class="ktc-pop-up-item-icon"
        >
          <i
            aria-hidden="true"
            class="icon-test"
          />
        </div>
        <div
          class="ktc-pop-up-item-label"
        >
          Item4
        </div>
      </li>
      <li
        class="ktc-pop-up-item ktc-pop-up-item-double-column"
        data-component-identifier="GUID5"
        title="Item5
Item5"
      >
        <div
          class="ktc-pop-up-item-icon"
        >
          <i
            aria-hidden="true"
            class="icon-test"
          />
        </div>
        <div
          class="ktc-pop-up-item-label"
        >
          Item5
        </div>
      </li>
    </ul>
  </div>
</kentico-pop-up-listing>
`;

exports[`pop-up-listing rendering should render with no items available message 1`] = `
<kentico-pop-up-listing
  class="hydrated"
>
  
  <div
    class="ktc-pop-up-item-listing ktc-pop-up-item-listing--empty"
  >
    TEST-MESSAGE
  </div>
</kentico-pop-up-listing>
`;
